Implement derived diagrams

mirodriguezgal opened the pull request:
Implement secondary diagrams such as the Phylogenetic Trees, the Chord Diagram, the Radar Chart and the Genomic Diagrams.

mirodriguezgal committed:
feat: Implement tree plots in plotly with hovertext features.
doc: Document tree usage at phylogenetic_tree.ipynb and plot_api_example.ipynb.
feat: Implement parse_color for deriving custom continous color scales from css strings.

mirodriguezgal committed:
feat : Implement genomic_features at the Track level.

mirodriguezgal committed:
fix: solve array length at calc_gc_content and  calc_gc_skew.

Now creates positions from 0 to len(seq) in step_size increments, then explicitly adds len(seq) if needed.

Old version potentially creates an extra position by going to len(seq)+step_size.

mirodriguezgal committed:
feat: Implement hovertext for genomic_features.

This substututes the matplotlib-based function of annotate.

mirodriguezgal committed:
feat: Incorporate hovertext in add_cytoband at Circos level.

fix: Change default radius for text at Sector level (sector name).

mirodriguezgal committed:
feat: Implement in ColorCycler the use of sequential palettes.

feat: hover_text could be disabled in heatmap at the Track level.

mirodriguezgal committed:
feat: Implement chord diagram.

fix: Autodetect type of palette at set_palette in ColorCycler.

mirodriguezgal committed:
fix: Hovertext background color in circos_plot.ipynb docs.

Without specifying fillcolor, it was getting grey as the default for links.

mirodriguezgal committed:
feat: parse_color now can deal with hex codes with opacity.

doc: Implement plotly-based documentation for comparative_genomics.ipynb.

mirodriguezgal committed:
feat: Implement Radar Chart

refactor: Reorganize default aesthetics and remove redundant and verlapping ones.

feat: All hovertexts in Tracks could be disabled by None.

mirodriguezgal committed:
feat: Implement stacked_bar and stacked_barh at the Track level. Includes default hovertext for stacked bars.

doc: Document plot_tips.ipynb

chore: Reorganize package to avoid circular imports. Moved color utils from helper.py to new color.py

mirodriguezgal committed:
feat: Default hover_text now rounds positional values to a precision level, by default defined by order of magnitude of track size.

gnuille commented on src/pycirclizely/circos.py:
```diff
@@ -797,3 +895,13 @@ def _get_all_traces(self) -> list[BaseTraceType]:
+        all_treeviz_list : list[TreeViz]
+            All tree visualization instance list
+        """
+        return list(itertools.chain(*[t._trees for t in self.tracks]))
```
I think with this you may not need to create an intermediate list
```python
list(itertools.chain.from_iterable(t._trees for t in self.tracks))
```
For reference: [https://www.geeksforgeeks.org/python/python-itertools-chain-from_iterable/](https://www.geeksforgeeks.org/python/python-itertools-chain-from_iterable/)

mirodriguezgal replied:
Yes, I applied the changes there and in other places with the same casuistic.

gnuille commented on src/pycirclizely/parser/genbank.py:
```diff
@@ -209,21 +209,25 @@ def calc_gc_content(
-        pos_list = np.append(pos_list, len(seq))
-
-        gc_content_list = np.zeros_like(pos_list, dtype=np.float64)
+        positions: list[int] = list(range(0, len(seq), step_size))
```
Is explicit typing needed? 
```python
        positions = list(range(0, len(seq), step_size))
```

mirodriguezgal replied:
Now, not really. Removed.

gnuille commented on src/pycirclizely/parser/genbank.py:
```diff
@@ -209,21 +209,25 @@ def calc_gc_content(
+        positions: list[int] = list(range(0, len(seq), step_size))
+        if not positions or positions[-1] != len(seq):
+            positions.append(len(seq))
+        pos_list: NDArray[np.int64] = np.array(positions, dtype=np.int64)
```
again typing probably unecessary

mirodriguezgal replied:
Done

gnuille commented on src/pycirclizely/parser/genbank.py:
```diff
@@ -209,21 +209,25 @@ def calc_gc_content(
+        pos_list: NDArray[np.int64] = np.array(positions, dtype=np.int64)
 
+        # Initialize and fill GC content array
+        gc_content_list: NDArray[np.float64] = np.empty_like(pos_list, dtype=np.float64)
```
unecessary typing

mirodriguezgal replied:
Removed.

gnuille commented on src/pycirclizely/track.py:
```diff
@@ -1671,6 +1625,20 @@ def genomic_features(
+                ),
+                hoverlabel={"bgcolor": bg_color},
+            )
+            self._traces.append(hover_trace)
```
Can we create private functions for both hover code? Or even If you feel powerful, a small contained class in a separate file with the name HoverFormatter that gets initialized at the beginning of the function and then calls its methods for both the first and second if in the function body. The initialization of the hover state (variables hover_texts, hover_x and hover_y could be moved inside the class and initializated on the init).

mirodriguezgal replied:
Done. In fact I also generated a types file to make it more readable.

gnuille commented on src/pycirclizely/circos.py:
```diff
@@ -313,18 +314,58 @@ def add_cytoband_tracks(
+                    if show_hovertext:
+                        # Calculate midpoint for hover point
+                        midpoint = (rec.start + rec.end) / 2
+                        rad = track.x_to_rad(midpoint)
```
Are we repeating code from previous hover in track?
Thats why creating a class could allow you to avoid repeated code

mirodriguezgal replied:
There are computations specific to the type of track, like computations of midpoints. However, now the hovertext commonalities are enclosed in the class.

gnuille commented on src/pycirclizely/track.py:
```diff
@@ -1334,9 +1334,9 @@ def heatmap(
-        hover_text : list[str] | None, optional
-            Custom hover text for each bar.
-            If None, defaults to formatted range x and height.
+        hover_text: list[str] | Literal["default"] | None = "default",
```
`        hover_text : list[str] | Literal["default"] | None = "default",`

gnuille replied:
the space before the semicolon is removed (i think) accidentaly

dpinol commented on .pre-commit-config.yaml:
```diff
@@ -74,11 +76,14 @@ repos:
           - pytz==2025.2
           - six==1.17.*
           - tzdata==2025.2
+          - webcolors
```
please pin at least minor version

mirodriguezgal replied:
Done

gnuille commented on docs/circos_plot.ipynb:
Please to avoid all this noise in changes between notebooks use `pre-commit` and `nb-clean` as we have in `sc-api` to force notebook cleaning and avoid spamming in all commits with huge changes

mirodriguezgal replied:
I skipped it because for the docs I need pre-computed plots. I will include nb-clean and make the docs rerun everything.

dpinol commented on src/pycirclizely/circos.py:
```diff
@@ -150,6 +157,458 @@ def tracks(self) -> list[Track]:
+
+        Parameters
+        ----------
+        table : str | Path | pd.DataFrame | RadarTable
```
please remove types as they're already documented in function type hints

mirodriguezgal replied:
Done. I simplified docs a lot and used Args scapi format.

dpinol commented on src/pycirclizely/circos.py:
```diff
@@ -150,6 +157,458 @@ def tracks(self) -> list[Track]:
+        Returns
+        -------
+        circos : Circos
+            Circos instance initialized for radar chart
```
let's avoid bureaucracy documentation. I'd remove the return section since it does not add value

mirodriguezgal replied:
Agree. I was just keeping the same docs as the original package behind the fork.

dpinol commented on src/pycirclizely/circos.py:
```diff
@@ -150,6 +157,458 @@ def tracks(self) -> list[Track]:
+        circos : Circos
+            Circos instance initialized for radar chart
+        """
+        if not vmin < vmax:
```
this function is too long in a too long file. I suggest moving the code to  a new impl/radar.py file which will have the implementation but not the docstring. In circos.py you can leave the function with doc, which just calls the function at impl/radar.py

 In the future it should be split into smaller functions, but that's a job for copilot

mirodriguezgal replied:
Done.

dpinol commented on src/pycirclizely/circos.py:
```diff
@@ -150,6 +157,458 @@ def tracks(self) -> list[Track]:
+            Circos instance initialized for radar chart
+        """
+        if not vmin < vmax:
+            raise ValueError(f"vmax must be larger than vmin ({vmin=}, {vmax=})")
```
try adding this to .pre-commit-config.yaml to detect duplicated code as Guillem mentioned
```
  - repo: https://github.com/jendrikseipp/vulture
    # dead code detection. Remove when https://github.com/astral-sh/ruff/issues/872 implemented
    rev: v2.14
    hooks:
      - id: vulture
        args: [--min-confidence, '100']
```

dpinol commented on src/pycirclizely/circos.py:
```diff
@@ -150,6 +157,458 @@ def tracks(self) -> list[Track]:
+        link_kws: dict[str, Any] | None = None,
+        link_kws_handler: Callable[[str, str], dict[str, Any] | None] | None = None,
+    ) -> Circos:
+        """Plot chord diagram
```
same as before. avoid duplicated doc & move to a new file

mirodriguezgal replied:
Done

dpinol commented on src/pycirclizely/circos.py:
```diff
@@ -150,6 +157,458 @@ def tracks(self) -> list[Track]:
+        label_formatter: Callable[[str], str] | None = None,
+        align_line_kws: dict[str, Any] | None = None,
+    ) -> tuple[Circos, TreeViz]:
+        """Initialize Circos instance from phylogenetic tree
```
same

dpinol commented on src/pycirclizely/circos.py:
```diff
@@ -797,3 +1299,13 @@ def _get_all_traces(self) -> list[BaseTraceType]:
+    def _get_all_treeviz_list(self) -> list[TreeViz]:
+        """Get all tree visualization instance list from tracks
+
+        Returns
+        -------
+        all_treeviz_list : list[TreeViz]
+            All tree visualization instance list
```
remove

dpinol commented on src/pycirclizely/track.py:
```diff
@@ -886,8 +835,9 @@ def scatter(
         hover_text : list[str] | None, optional
-            Custom hover text for each point in line.
-            If None, defaults to formatted x and y values.
+            - "default": Auto-generates hover text (default)
```
To avoid repeating the types and documentation, you have 2 alternatives:
1) create a class HoverText with implementation (best option if possible)
2) `HoverText: TypeAlias = list[str] | None, optional` which will not centralize implementation , but at least centralizes the types and documentation

dpinol commented on src/pycirclizely/track.py:
```diff
@@ -1386,284 +1483,321 @@ def heatmap(
+
+        Parameters
+        ----------
+        tree_data : str | Path | Tree
```
remove duplicated types

dpinol commented on src/pycirclizely/track.py:
```diff
@@ -1386,284 +1483,321 @@ def heatmap(
+        Returns
+        -------
+        tv : TreeViz
+            TreeViz instance
```
remove

dpinol commented on src/pycirclizely/track.py:
```diff
@@ -1386,284 +1483,321 @@ def heatmap(
+
+        return tv
+
+    def genomic_features(
```
why does function "tree" return TreeViz but genomic_features doesn't?

mirodriguezgal replied:
Genomic_features add Shapes, annotations and traces to the parameters of the class 
        self._shapes: list[go.layout.Shape] = []
        self._annotations: list[go.layout.Annotation] = []
        self._traces: list[BaseTraceType] = []
which are called at the time of plotting the figure. Usually the plotting functions do not have a return, but tree in fact initializes a tree class with some parameters and return the class, which can be modified for aesthethics.

The way of doing this is the same as the original pyCirclize package, and I believe it is a good practice.

dpinol commented on src/pycirclizely/utils/color.py:
```diff
@@ -0,0 +1,207 @@
+
+    @classmethod
+    def get_color(cls, n: Optional[int] = None) -> str:
+        """Get a color from the current palette."""
```
can you document how "n" works and its relationship with reset_cycle?
Maybe provide overview in the class docstring?

mirodriguezgal replied:
Done

dpinol commented on src/pycirclizely/utils/color.py:
```diff
@@ -0,0 +1,207 @@
+            return colors[n % len(colors)]
+
+    @classmethod
+    def get_color_list(cls, n: Optional[int] = None) -> List[str]:
```
this n I guess is very different from the n argument in get_color. I would choose different names

mirodriguezgal replied:
Fair enough. See the entire class refactoring.

dpinol commented on src/pycirclizely/utils/color.py:
```diff
@@ -0,0 +1,207 @@
+
+    @classmethod
+    def set_palette(cls, name: str, palette_type: Optional[str] = None) -> None:
+        """Set the current color palette by name.
```
try to avoid global "singleton" configuration. It would be more robust to have to create one instance of ColorCycler per each different configuration. Would it be too complex to redesign it?

mirodriguezgal replied:
Done

dpinol commented on src/pycirclizely/utils/color.py:
```diff
@@ -0,0 +1,207 @@
+        return cls._palette_type[cls._current_palette]
+
+
+def parse_color(color):
```
type argument

mirodriguezgal committed:
git filter-repo: Remove docs/circos_plot.ipynb from history. All previous commits squashed to solve history problems:

refactor: Move radar_chart and coord_diagram to private files.

chore: Add nb-clean at pre-commit pipeline.

refactor: Hovertext is managed by a common class (except tree hovertexts).

doc: Update with only essential docstring.

dpinol commented on src/pycirclizely/types.py:
```diff
@@ -0,0 +1,21 @@
+
+import numpy as np
+
+Numeric = Union[int, float]
```
Numeric: TypeAlias = Union[int, float]

dpinol commented on src/pycirclizely/utils/color.py:
```diff
@@ -1,169 +1,178 @@
+class ColorPalette:
+    """Container for color palette configuration."""
+
+    class PaletteType(Enum):
```
Consider renaming to "Type" to avoid "stuttering": ColorPalette.PaletteType

dpinol commented on src/pycirclizely/utils/color.py:
```diff
@@ -1,169 +1,178 @@
+class ColorPalette:
+    """Container for color palette configuration."""
+
+    class PaletteType(Enum):
```
Use StrEnum. If you get any mypy or similar error (happens rarely) just leave as Enum

mirodriguezgal replied:
Problem is that it is not available in python 3.10. I'll leave it as Enum so it works from python 3.10 to 3.13.

dpinol replied:
:+1:

dpinol commented on src/pycirclizely/utils/hover_formatter.py:
```diff
@@ -0,0 +1,212 @@
+
+
+class HoverFormatter:
+    hover_texts: List[str]
+    hover_x: List[float]
+    hover_y: List[float]
+    hover_colors: List[str]
```
unless other languages, in python this syntax declares *static fields" instead of instance fields. You have 2 options:
1) convert to dataclass. Then you can use this syntax
2) remove and move field declaration to assignment in constructor

dpinol commented on src/pycirclizely/utils/hover_formatter.py:
```diff
@@ -0,0 +1,212 @@
+
+    def clear(self) -> None:
+        """Clear all stored hover data."""
+        self.hover_texts = []
```
I've seen cases where an external class gets a reference to a list field, and then after clear you end up with duplicated list instances.
To avoid this, do self.hover_texts.clear()

dpinol commented on src/pycirclizely/parser/table.py:
```diff
@@ -15,24 +15,35 @@ def __init__(
         """Table dataframe"""
         return self._dataframe
 
+    @property
+    def color_cycler(self) -> ColorCycler:
+        """Get the color cycler instance"""
+        return self._color_cycler
+
+    @color_cycler.setter
+    def color_cycler(self, cycler: ColorCycler):
+        """Set the color cycler instance"""
+        self._color_cycler = cycler
```
this looks like has no effect. Consider removing it and renaming _color_cycler to color_cycler

mirodriguezgal committed:
fix: Add default ColorCycler instance so get_default_color does not retrieve always the same default.

mirodriguezgal committed:
refactor: improve type hints and class implementations

- refactor(type): use TypeAlias type on types
- fix(HoverFormatter): convert to dataclass to fix instance field declaration
- fix(HoverFormatter): use list.clear() instead of reassignment
- refactor(Table): remove unnecessary property wrapper for color_cycler in Table parser class

gnuille commented on .pre-commit-config.yaml:
```diff
@@ -85,6 +85,13 @@ repos:
+    hooks:
+      # https://github.com/srstevenson/nb-clean/issues/323
+      - id: nb-clean
+        entry: nb-clean check --preserve-execution-counts --preserve-notebook-metadata
```
why preserve execution counts?

mirodriguezgal replied:
I used the sc-api config. Honestly, I do not care about execution counts nor metadata. If there is no reason, maybe changing at sc-api is agood idea.

gnuille commented on src/pycirclizely/utils/hover_formatter.py:
```diff
@@ -0,0 +1,215 @@
+        self.hover_texts = []
+        self.hover_x = []
+        self.hover_y = []
+        self.hover_colors = []
```
dataclass and explicit constructor is a weird pattern, you can remove the `__init__` constructor and leave default values for the dataclass:
```python
@dataclass
class HoverFormatter:
    precision_position: int = 2
    hover_texts: List[str] = field(default_factory=list)
    hover_x: List[float] = field(default_factory=list)
    hover_y: List[float] = field(default_factory=list)
    hover_colors: List[str] = field(default_factory=list)

gnuille replied:
Also we don't use `typing.List`  we use builtin `list`
```python
@dataclass
class HoverFormatter:
    precision_position: int = 2
    hover_texts: list[str] = field(default_factory=list)
    hover_x: list[float] = field(default_factory=list)
    hover_y: list[float] = field(default_factory=list)
    hover_colors: list[str] = field(default_factory=list)
```

mirodriguezgal committed:
refactor(ColorPalette): rename PaletteType to Type to avoid stuttering.

refactor(HoverFormatter): Remove explicit __init__ at since dataclass handles initialization.

chore: Maintain existing functionality while fixing type checking.
